numpy/core/src/umath/ufunc_object.c

```diff
@@ -5178,7 +5178,7 @@ PyUFunc_RegisterLoopForDescr(PyUFuncObject *ufunc,
                 }
                 current = current->next;
             }
-            if (cmp == 0 && current->arg_dtypes == NULL) {
+            if (cmp == 0 && current != NULL && current->arg_dtypes == NULL) {
```

Comment:
At least from the code here, it looks like one can in principle have had the last in the chain have `cmp == 0` but `last->arg_dtypes != NULL`.

Total nitpick: could you put `current != NULL` at the start?  Or perhaps this is clearer if inverted
```
if (current == NULL || !(cmp == 0 && current->arg_dtypes == NULL)) {
    result = -1;
}
else {
    ...
}
```

Reply:
This whole function looks to be in a poor state to me, and will return `-1` without setting an exception.

Reply:
> will return `-1` without setting an exception

I noticed that too, but in the few cases the `PyUFunc_RegisterLoopForDescr` is used there's a if `(!PyErr_Occurred()) { PyErr_SetString(PyExc_RuntimeError...` or the output is not checked at all.

Reply:
added a test that reaches the "return -1", added exception string